jina/executors/crafters/nlp/split.py

```diff
@@ -97,3 +99,73 @@ def craft(self, buffer: bytes, doc_id: int, *args, **kwargs) -> List[Dict]:
                 dict(text=word, offset=idx, weight=1.0))
 
         return chunks
+
+
+class SlidingWindowSegmenter(BaseSegmenter):
+    """
+    :class:`SlidingWindowSegmenter` split the text on the doc-level into overlapping substrings on the chunk-level.
+        The text is split into substrings of length ``window_size`` if possible.
+        The degree of overlapping can be configured through the ``step_size`` parameter.
+        The substrings that are shorter than the ``min_substring_len`` will be discarded.
+    """
+
+    def __init__(self,
+                 window_size: int = 300,
+                 step_size: int = 150,
+                 min_substring_len: int = 1,
+                 *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.window_size = window_size
+        self.step_size = step_size
+        self.min_substring_len = min_substring_len
+        if self.min_substring_len > self.window_size:
+            self.logger.warning(
+                'the min_substring_len (={}) should be smaller to the window_size (={})'.format(
+                    self.min_substring_len, self.window_size))
+        if self.window_size <= 0:
+            self.logger.warning(
+                'the window_size (={}) should be larger than zero'.format(
+                    self.window_size))
+        if self.step_size > self.window_size:
+            self.logger.warning(
+                'the step_size (={}) should not be larger than the window_size (={})'.format(
+                    self.window_size, self.step_size))
+
+    def craft(self, buffer: bytes, doc_id: int, *args, **kwargs) -> List[Dict]:
+        """
+        Split the text into overlapping chunks
+        :param buffer: the raw text in the `bytes` format
+        :param doc_id: the doc id
+        :return: a list of chunk dicts
+        """
+
+        def sliding_window(iterable, size, step):
+            i = iter(text)
+            d = deque(islice(i, size),
+                      maxlen=size)
+            if not d:
+                # empty text
+                return results
+            while True:
+                yield iter(d)
+                try:
+                    d.append(next(i))
+                except StopIteration:
+                    return
+                d.extend(next(i, None)
+                         for _ in range(step-1))
+
+        text = buffer.decode('utf-8')
+        chunks = ["".join(filter(None, list(chunk))) for chunk in sliding_window(
```

Comment:
single quote is the preferred coding style of this project

```suggestion
        chunks = [''.join(filter(None, list(chunk))) for chunk in sliding_window(
```